File: harmony/push_notification/src/main/ets/PushNotificationTurboModule.ts
// Copyright (c) 2024 Huawei Device Co., Ltd. All rights reserved
// Use of this source code is governed by a MIT license that can be
// found in the LICENSE file.

import { TurboModule } from '@rnoh/react-native-openharmony/ts';
import notificationManager from '@ohos.notificationManager';
import { NotificationRequest } from './NotificationRequest';
import { ConvertUtils } from './ConvertUtils';
import Logger from './Logger';

export class PushNotificationTurboModule extends TurboModule {
  private TAG: string = "PushNotificationTurboModule";
  utils: ConvertUtils = new ConvertUtils();

  addNotificationRequest(request: NotificationRequest) {
    this.utils.convertNotification(request).then((data) => {
      if (data) {
        notificationManager.publish(data, (err) => {
          if (err) {
            Logger.info(this.TAG, `publish failed, code is ${err.code}, message is ${err.message}`);
          } else {
            Logger.info(this.TAG, "publish success");
          }
        });
      } else {
        Logger.info(this.TAG, "publish failed");
      }
    })
  }

  removeDeliveredNotifications(identifiers: Array<string>) {
    for (let i = 0; i < identifiers.length; i++) {
      let data = identifiers[i];
      let id = this.utils.parsePositiveInteger(data);

      notificationManager.cancel(id, (err) => {
        if (err) {
          Logger.info(this.TAG, `cancel failed, code is ${err.code}, message is ${err.message}`);
        } else {
          Logger.info(this.TAG, "cancel success");
        }
      })
    }
  }

  removeAllDeliveredNotifications() {
    notificationManager.cancelAll((err) => {
      if (err) {
        Logger.info(this.TAG, `cancelAll failed, code is ${err.code}, message is ${err.message}`);
      } else {
        Logger.info(this.TAG, "cancelAll success");
      }
    });
  }

  setApplicationIconBadgeNumber(badgeNumber: number) {
    notificationManager.setBadgeNumber(badgeNumber, (err) => {
      if (err) {
        Logger.info(this.TAG, `setBadge fail: ${JSON.stringify(err)}`);
      } else {
        Logger.info(this.TAG, "setBadge success");
      }
    })
  }

  getDeliveredNotifications(callback: (result?: Array<object>) => void) {
    notificationManager.getActiveNotifications().then((data) => {
      let newArray = this.utils.transformArray(data)
      callback(newArray);
    })
  }
}

File: harmony/push_notification/src/main/cpp/PushNotificationTurboModule.cpp
// Copyright (c) 2024 Huawei Device Co., Ltd. All rights reserved
// Use of this source code is governed by a MIT license that can be
// found in the LICENSE file.

#include "PushNotificationTurboModule.h"
#include "RNOH/ArkTSTurboModule.h"

using namespace rnoh;
using namespace facebook;

static jsi::Value __hostFunction_PushNotificationTurboModule_removeAllDeliveredNotifications(
    jsi::Runtime &rt, 
    react::TurboModule &turboModule, 
    const jsi::Value *args, 
    size_t count) {
    return jsi::Value(static_cast<ArkTSTurboModule &>(turboModule).call(rt, "removeAllDeliveredNotifications", args, count));
}

static jsi::Value __hostFunction_PushNotificationTurboModule_removeDeliveredNotifications(
    jsi::Runtime &rt,
    react::TurboModule &turboModule,
    const jsi::Value *args,
    size_t count) {
    return jsi::Value(static_cast<ArkTSTurboModule &>(turboModule).call(rt, "removeDeliveredNotifications", args, count));
}

static jsi::Value __hostFunction_PushNotificationTurboModule_addNotificationRequest(
    jsi::Runtime &rt,
    react::TurboModule &turboModule,
    const jsi::Value *args,
    size_t count) {
    return jsi::Value(static_cast<ArkTSTurboModule &>(turboModule).call(rt, "addNotificationRequest", args, count));
}

static jsi::Value __hostFunction_PushNotificationTurboModule_setApplicationIconBadgeNumber(
    jsi::Runtime &rt,
    react::TurboModule &turboModule,
    const jsi::Value *args,
    size_t count) {
    return jsi::Value(static_cast<ArkTSTurboModule &>(turboModule).call(rt, "setApplicationIconBadgeNumber", args, count));
}

static jsi::Value __hostFunction_PushNotificationTurboModule_getDeliveredNotifications(
    jsi::Runtime &rt,
    react::TurboModule &turboModule,
    const jsi::Value *args,
    size_t count) {
    return jsi::Value(static_cast<ArkTSTurboModule &>(turboModule).call(rt, "getDeliveredNotifications", args, count));
}

PushNotificationTurboModule::PushNotificationTurboModule(const ArkTSTurboModule::Context ctx, const std::string name)
  : ArkTSTurboModule(ctx, name) {
  methodMap_["removeAllDeliveredNotifications"] = MethodMetadata {0, __hostFunction_PushNotificationTurboModule_removeAllDeliveredNotifications};
  methodMap_["removeDeliveredNotifications"] = MethodMetadata {1, __hostFunction_PushNotificationTurboModule_removeDeliveredNotifications};
  methodMap_["addNotificationRequest"] = MethodMetadata {1, __hostFunction_PushNotificationTurboModule_addNotificationRequest};
  methodMap_["setApplicationIconBadgeNumber"] = MethodMetadata {1, __hostFunction_PushNotificationTurboModule_setApplicationIconBadgeNumber};
  methodMap_["getDeliveredNotifications"] = MethodMetadata {1, __hostFunction_PushNotificationTurboModule_getDeliveredNotifications};
}
